validate_sptx/util.py

```diff
@@ -107,16 +114,202 @@ def validate_object(self, target_object: Dict, target_file: str=None) -> bool:
+        """
+        header = f"{self.state()}"
+        header += "If the letter is present, mutation is valid!"
+        self.out.write(f"{header}\n")
+        self.out.write("".join([x in ("\t", "\n") and x or "-" for x in header]))
+        self.out.write("\n")
+        self.stack : List[Any] = []
+        try:
+            self.descend(self.obj)
+        finally:
+            self.stack = None
+
+    def state(self) -> str:
+        """primary driver for the checks of individual trees
+
+        Returns
+        -------
+        str :
+            space-separated representation of the fuzzing conditions.
+            If a letter is present, then mutation leaves the tree in
+            a valid state:
+
+             A: inserting a fake key or appending to a list
+             D: deleting a key or index
+             I: converting value to an integer
+             I: converting value to a string
+             M: converting value to an empty dict
+             L: converting value to an empty list
+
+        """
+        rv = [
+            Add().check(self),
+            Del().check(self),
+            Change("I", lambda *args: 123456789).check(self),
+            Change("S", lambda *args: "fake").check(self),
+            Change("M", lambda *args: dict()).check(self),
+            Change("L", lambda *args: list()).check(self),
+        ]
+        return ' '.join(rv) + "\t"
+
+    def descend(self, obj : Any, depth : int=0, prefix : str="") -> None:
+        """walk a JSON-like object tree printing the state of the tree
+        at each level. A YAML representation is used for simplicity.
+
+        Parameters
+        ----------
+        obj : Any
+            JSON-like object tree
+        depth : int
+            depth in the tree that is currently being evaluated
+        prefix : str
+            value which should be prepended to printouts at this level
+        """
+        if self.stack is None:
+            raise Exception("invalid state")
+        if isinstance(obj, list):
+            for i, o in enumerate(obj):
+                depth += 1
+                self.stack.append(i)
+                self.descend(o, depth, prefix="- ")
+                self.stack.pop()
+                depth -= 1
+        elif isinstance(obj, dict):
+            for k in obj:
+                # This is something of a workaround in that we need a special
+                # case for object keys since no __getitem__ method will suffice.
+                self.stack.append((k,))
+                self.out.write(f"{self.state()}{' ' * depth}{prefix}{k}:\n")
+                self.stack.pop()
+                if prefix == "- ": prefix = "  "
+                depth += 1
+                self.stack.append(k)
+                self.descend(obj[k], depth, prefix="  "+prefix)
+                self.stack.pop()
+                depth -= 1
+        else:
+            self.out.write(f"{self.state()}{' ' * depth}{prefix}{obj}\n")
+
+class Checker(object):
```

Comment:
Is this class designed to be abstract? Should `LETTER` be `None` or a `property` that raises a `NotImplemented` error?

Reply:
Designed to be abstract. There are several different strategies in Python land. I don't see one using `@property` & `NotImplemented` in the codebase. Happy to follow whatever pattern is appropriate.

Reply:
Yea, this is perhaps an @ttung and @shanaxel42 question about what they prefer. 

I like the clarity of simple assignments like what you've done, although I'd assign it to `None` with a note. 

```python 
LETTER = None  # sub-classes must set this property to an upper-case string letter
```

However, an `@property` that raises `NotImplementedError` actually enforces the contract: 

```python
@property
def LETTER(self) -> str:
    raise NotImplementedError
```

Reply:
Maybe clearer is `LETTER: Optional[str] = None  # ... `

Reply:
`validate_sptx/util.py:270: error: Incompatible return value type (got "Optional[str]", expected "str")` for the latter.

Reply:
Yea... mypy doesn't play well with abstract stuff. :(

Reply:
I like the `@property` suggestion.  (good thing there isn't someone with the u/n property...)